cumulus_library/studies/core/core_templates/core_utils.jinja

```diff
@@ -1,38 +1,138 @@
+    tuple(col, object, , child, col_alias): refers to a two level nested SQL object
 
--- gets a column, if present, from a basic SQL primitive col, or a null otherwise
-{% macro basic_nullable_cols(table, alias, targets, schema) %}
+-#}
+{% macro nullable_cols(table, alias, targets, schema) %}
         {%- for col in targets %}
-        {%- if schema[table][col] %}
+        {#- depth one nested column-#}
+        {%- if col is not string and col|length ==3%}
+        {%- if schema[table][col[0]][col[1]] %}
+        {{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }} AS {{col[2]}}
+        {%- else %}
+        NULL AS {{col[2]}}
+        {%- endif %}
+        {#- depth two nested column -#}
+        {%- elif col is not string and col|length ==4%}
+        {%- if schema[table][col[0]][col[2]] %}
+        {{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }}.{{ col[2].lower() }} AS {{col[3]}}
+        {%- else %}
+        NULL AS {{col[3]}}
+        {%- endif %}
+        {#- SQL primitive column column-#}
+        {%- elif schema[table][col] %}
         {{ alias }}.{{ col.lower() }}
         {%- else %}
         NULL AS {{ col.lower() }}
+        {%- endif %}        
+        {{- comma_delineate(loop) }}
+        {%- endfor %}
+{%- endmacro %}
+
+{#- creates a date from a from an ISO formatted date str 
+targets is an array expecting a data type of the following:
+    string: refers to a vanilla SQL primitive column
+    tuple(col, object, col_alias): refers to a one level nested SQL object
+    tuple(col, object, , child, col_alias): refers to a two level nested SQL obj
+-#}
+{% macro date_cols_from_str(table, alias, targets, schema) %}
+        {%- for col in targets %}
+        {#- depth one nested column-#}
+        {%- if col is not string and col|length ==3%}
+        {%- if schema[table][col[0]][col[1]] %}
+        date(from_iso8601_timestamp({{ alias }}.{{ col[0] }}.{{ col[1].lower() }})) AS {{ col[2].lower() }}
+        {%- else %}
+        NULL AS {{col[1]}}
+        {% endif %}
+        {#- depth two nested column -#}
+        {%- elif col is not string and col|length ==4%}
+        {%- if schema[table][col[0]][col[2]] %}
+        date(from_iso8601_timestamp({{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }}.{{ col[2].lower() }})) AS {{col[3]}}
+        {%- else %}
+        NULL AS {{col[3]}}
         {%- endif %}
+        {#- SQL primitive column column-#}
+        {%- elif schema[table][col] %}
+        date(from_iso8601_timestamp({{ alias }}.{{ col.lower() }})) AS {{ col.lower() }}
+        {%- else %}
+        cast(NULL AS date) AS {{ col.lower() }}
+        {%- endif %}    
         {{- comma_delineate(loop) }}
         {%- endfor %}
 {%- endmacro %}
 
--- gets a column, if present, from a value of an object in a col
-{% macro depth_one_nullable_cols(table, alias, targets, schema) %}
+{#- creates a date from a list of columns that may contain a partial date str -#}
+{% macro partial_date_cols(table, alias, targets, schema) %}
```

Comment:
Why does this need to exist? (i.e. why not just call `date_cols_from_str`)  I believe this function's logic is already the behavior of `from_iso8601_timestamp` - and I coded this logic into the duckdb version of that function too.

Reply:
mostly for the schema validation step - but the case block could be replaced with `date_cols_from string`.

Reply:
`date_cols_from_string` also checks the schema, yeah?

Reply:
yep - removed